### PR TITLE
[IMP] pivot: allow to sort the pivot on any column

### PR DESCRIPTION
**Pull request by hokolomopo**
## Description

This commit allows the user to right click on a pivot cell to sort the pivot on that column.

Task: [3989395](https://www.odoo.com/web#id=3989395&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_t("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/helpers/pivot/pivot_domain_helpers.ts [hunk 1]
```diff
@@ -253,3 +253,27 @@ export function getRunningTotalDomainKey(
...
+      [...domain, { field: node2.field, value: node2.value, type: node2.type }]
+    )
+  );
+  for (const node of tree) {
+    const children = sortPivotTree(
+      node.children,
+      [...domain, { field: node.field, value: node.value, type: node.type }],
+      sortFn
+    );
+    node.children = children;
```
Comment (VincentSchippefilt) [inline src/helpers/pivot/pivot_domain_helpers.ts:276]: do we need this many copies of arrays ?

Comment (hokolomopo) [reply src/helpers/pivot/pivot_domain_helpers.ts]: I can definetly imrpove to create only one array (the domain) per tree node 🙃

File: src/components/side_panel/pivot/pivot_side_panel/pivot_side_panel_store.ts [hunk 2]
```diff
@@ -195,6 +195,7 @@ export class PivotSidePanelStore extends SpreadsheetStore {
         format: measure.format,
         display: measure.display,
       })),
+      sortedCol: this.shouldRemoveSortedColumn(definition) ? definition.sortedCol : undefined,
```
Comment (LucasLefevre) [inline src/components/side_panel/pivot/pivot_side_panel/pivot_side_panel_store.ts:198]: The name of `shouldRemoveSortedColumn` is a bit misleading. This line reads as "If the sorted columns should be removed, then don't remove it"

File: src/helpers/pivot/pivot_domain_helpers.ts [hunk 3]
```diff
@@ -253,3 +253,27 @@ export function getRunningTotalDomainKey(
...
+  tree: DimensionTree,
+  baseDomain: PivotDomain,
+  sortFn: (a: PivotDomain, b: PivotDomain) => number
+) {
+  const sortedTree = [...tree];
+  const domain = [...baseDomain];
+  sortedTree.sort((node1, node2) =>
+    sortFn(
+      [...domain, { field: node1.field, value: node1.value, type: node1.type }],
+      [...domain, { field: node2.field, value: node2.value, type: node2.type }]
```
Comment (LucasLefevre) [inline src/helpers/pivot/pivot_domain_helpers.ts:267]: nitpick, we can avoid creating new objects as both types are compatible
```suggestion
      [...domain, node1],
      [...domain, node2]
```

File: src/helpers/pivot/pivot_domain_helpers.ts [hunk 4]
```diff
@@ -253,3 +253,27 @@ export function getRunningTotalDomainKey(
...
+    )
+  );
+  for (const node of tree) {
+    const children = sortPivotTree(
+      node.children,
+      [...domain, { field: node.field, value: node.value, type: node.type }],
```
Comment (LucasLefevre) [inline src/helpers/pivot/pivot_domain_helpers.ts:273]:
```suggestion
      [...domain, node],
```

File: src/helpers/pivot/table_spreadsheet_pivot.ts [hunk 5]
```diff
@@ -305,6 +311,42 @@ export class SpreadsheetPivotTable {
...
+    };
+    const sortColDomain = sortedCol.domain;
+
+    const sortFn = (rowDomain1: PivotDomain, rowDomain2: PivotDomain) => {
+      const value1 = getSortValue(measure, [...rowDomain1, ...sortColDomain]);
+      const value2 = getSortValue(measure, [...rowDomain2, ...sortColDomain]);
+      return sortedCol.order === "asc" ? value1 - value2 : value2 - value1;
+    };
+    const sortedRowTree = sortPivotTree(this.rowTree(), [], sortFn);
+    this.rowTree = lazy(() => sortedRowTree);
```
Comment (LucasLefevre) [inline src/helpers/pivot/table_spreadsheet_pivot.ts:335]:
```suggestion
    this.rowTree = lazy(sortedRowTree);
```

File: src/helpers/pivot/table_spreadsheet_pivot.ts [hunk 6]
```diff
@@ -305,6 +311,42 @@ export class SpreadsheetPivotTable {
...
+    measure: string,
+    sortedCol: PivotSortedColumn,
+    getValue: (measure: string, domain: PivotDomain) => FunctionResultObject
+  ) {
+    if (this.isSorted) {
+      return;
+    }
+    const getSortValue = (measure: string, domain: PivotDomain): number => {
+      const rawValue = getValue(measure, domain).value;
+      return typeof rawValue === "number" ? rawValue : -Infinity; // the only non-number values are empty aggregates
```
Comment (LucasLefevre) [inline src/helpers/pivot/table_spreadsheet_pivot.ts:325]: I do not understand this comment :sweat_smile: Is the comment useful ? I feel the code is clear enough

File: src/helpers/pivot/table_spreadsheet_pivot.ts [hunk 7]
```diff
@@ -305,6 +311,42 @@ export class SpreadsheetPivotTable {
...
+    const sortColDomain = sortedCol.domain;
+
+    const sortFn = (rowDomain1: PivotDomain, rowDomain2: PivotDomain) => {
+      const value1 = getSortValue(measure, [...rowDomain1, ...sortColDomain]);
+      const value2 = getSortValue(measure, [...rowDomain2, ...sortColDomain]);
+      return sortedCol.order === "asc" ? value1 - value2 : value2 - value1;
+    };
+    const sortedRowTree = sortPivotTree(this.rowTree(), [], sortFn);
+    this.rowTree = lazy(() => sortedRowTree);
+    this.rows = [...this.rowTreeToRows(sortedRowTree), this.rows[this.rows.length - 1]]!;
```
Comment (LucasLefevre) [inline src/helpers/pivot/table_spreadsheet_pivot.ts:336]:
```suggestion
    this.rows = [...this.rowTreeToRows(sortedRowTree), this.rows[this.rows.length - 1]];
```

**Review by LucasLefevre — APPROVED** [review]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo rebase-ff

**Pull request closed without merging**